grails-app/services/org/pih/warehouse/product/ProductService.groovy

```diff
@@ -731,6 +731,18 @@ class ProductService {
             addTagsToProduct(product, tags)
             addTagsToProduct(product, productProperties.tags)
 
+            ProductType defaultProductType = ProductType.defaultProductType.list()?.first();
+            // Throw an error for product type with empty code and product identifier that is not a default product type
+            if (product.productType?.id != defaultProductType?.id && !product.productType?.code && !product.productType?.productIdentifierFormat) {
+                throw new ValidationException("Could not save product '" + product.name + "'", product.errors)
+            }
+
+            if (!product?.id || product.validate()) {
```

Comment:
I was just about to comment that we should probably check if the product is valid here before calling generateProductIdentifier ... and then I saw this. Good work.